iree/compiler/Conversion/LinalgToLLVMGPU/Common.h

```diff
@@ -0,0 +1,93 @@
+// Copyright 2021 Nod Labs
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IREE_COMPILER_CONVERSION_LINALGTOLLVMGPU_COMMON_H_
+#define IREE_COMPILER_CONVERSION_LINALGTOLLVMGPU_COMMON_H_
+
+#include "iree/compiler/Conversion/CodegenUtils/FunctionUtils.h"
+#include "iree/compiler/Dialect/IREE/IR/IREEOps.h"
+#include "mlir/Conversion/StandardToLLVM/ConvertStandardToLLVM.h"
+#include "mlir/Conversion/VectorToLLVM/ConvertVectorToLLVM.h"
+#include "mlir/Dialect/GPU/Passes.h"
+#include "mlir/Dialect/LLVMIR/NVVMDialect.h"
+#include "mlir/Dialect/LLVMIR/ROCDLDialect.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Dialect/Vector/VectorOps.h"
+#include "mlir/Transforms/Passes.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+class ConvertFunc : public ConvertToLLVMPattern {
+ public:
+  explicit ConvertFunc(MLIRContext *context, LLVMTypeConverter &converter)
+      : ConvertToLLVMPattern(mlir::FuncOp::getOperationName(), context,
+                             converter, 100) {}
+  LogicalResult matchAndRewrite(
+      Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override;
+};
+
+class ConvertIREEBindingOp : public ConvertToLLVMPattern {
+ public:
+  explicit ConvertIREEBindingOp(MLIRContext *context,
+                                LLVMTypeConverter &converter)
+      : ConvertToLLVMPattern(
+            IREE::HAL::InterfaceBindingSubspanOp::getOperationName(), context,
+            converter) {}
+  LogicalResult matchAndRewrite(
+      Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override;
+};
+
+/// A pattern to convert hal.interface.workgroup.id/count/size into
+/// corresponding NVVM/ROCDL ops.
+template <typename InterfaceOpTy, typename XOp, typename YOp, typename ZOp>
+struct HALInterfaceWorkgroupOpsConverter final
```

Comment:
We should lower to GPU dialect IDs, so that we don't need different behavior between NVVM and ROCDL. I think it is fine to do as one of the first step of ConvertToNVVM/ConvertToROCDL this way we don't have to keep this pattern in the header. Later we'll probably want to do it earlier.

Reply:
That's a good idea. Sure will keep this as is for now then. Should I create an issue for the lowering to GPU dialect IDs so that it is tracked as one of the next work items?